### ISSUE TITLE
Get exception Method not found: 'System.Collections.ObjectModel.Collection`1<System.Net.Http.DelegatingHandler> System.Web.Http.HttpConfiguration.get_MessageHandlers()' 

### ISSUE THREAD

**Issue by hemendpatel** [simpleinjector/SimpleInjector#566]
I am getting exception when trying to add 

``` c#
config.MessageHandlers.Add(
    new DelegatingHandlerProxy<RequestResponseLoggerHandler>());
```

config.MessageHandler working in my other project but as soon as I have added the Simple Injector reference and try to add the MessageHander to inject dependence it through the exception. I am using the same code provided in below link,

https://simpleinjector.readthedocs.io/en/2.6/webapiintegration.html
-->Injecting dependencies into Web API message handlers

**stack trace:**

[MissingMethodException: Method not found: 'System.Collections.ObjectModel.Collection`1<System.Net.Http.DelegatingHandler> System.Web.Http.HttpConfiguration.get_MessageHandlers()'.]    WebApplication2.WebApiConfig.Register(HttpConfiguration config) in WebApplication2\WebApplication2\App_Start\WebApiConfig.cs:41    System.Web.Http.GlobalConfiguration.Configure(Action`1 configurationCallback) +63    WebApplication2.WebApiApplication.Application_Start() in \WebApplication2\WebApplication2\Global.asax.cs:14  [HttpException (0x80004005): Method not found: 'System.Collections.ObjectModel.Collection`1<System.Net.Http.DelegatingHandler> System.Web.Http.HttpConfiguration.get_MessageHandlers()'.]    System.Web.HttpApplicationFactory.EnsureAppStartCalledForIntegratedMode(HttpContext context, HttpApplication app) +535    System.Web.HttpApplication.RegisterEventSubscriptionsWithIIS(IntPtr appContext, HttpContext context, MethodInfo[] handlers) +166    System.Web.HttpApplication.InitSpecial(HttpApplicationState state, MethodInfo[] handlers, IntPtr appContext, HttpContext context) +169    System.Web.HttpApplicationFactory.GetSpecialApplicationInstance(IntPtr appContext, HttpContext context) +372    System.Web.Hosting.PipelineRuntime.InitializeApplication(IntPtr appContext) +329  [HttpException (0x80004005): Method not found: 'System.Collections.ObjectModel.Collection`1<System.Net.Http.DelegatingHandler> System.Web.Http.HttpConfiguration.get_MessageHandlers()'.]    System.Web.HttpRuntime.FirstRequestInit(HttpContext context) +525    System.Web.HttpRuntime.EnsureFirstRequestInit(HttpContext context) +118    System.Web.HttpRuntime.ProcessRequestNotificationPrivate(IIS7WorkerRequest wr, HttpContext context) +708
--

**Comment by dotnetjunkie** [comment]
Please post the full stack trace.

**Comment by hemendpatel** [comment]
stack trace

[MissingMethodException: Method not found: 'System.Collections.ObjectModel.Collection`1<System.Net.Http.DelegatingHandler> System.Web.Http.HttpConfiguration.get_MessageHandlers()'.]    WebApplication2.WebApiConfig.Register(HttpConfiguration config) in C:\Users\x-hpatel\source\repos\WebApplication2\WebApplication2\App_Start\WebApiConfig.cs:41    System.Web.Http.GlobalConfiguration.Configure(Action`1 configurationCallback) +63    WebApplication2.WebApiApplication.Application_Start() in C:\Users\x-hpatel\source\repos\WebApplication2\WebApplication2\Global.asax.cs:14  [HttpException (0x80004005): Method not found: 'System.Collections.ObjectModel.Collection`1<System.Net.Http.DelegatingHandler> System.Web.Http.HttpConfiguration.get_MessageHandlers()'.]    System.Web.HttpApplicationFactory.EnsureAppStartCalledForIntegratedMode(HttpContext context, HttpApplication app) +535    System.Web.HttpApplication.RegisterEventSubscriptionsWithIIS(IntPtr appContext, HttpContext context, MethodInfo[] handlers) +166    System.Web.HttpApplication.InitSpecial(HttpApplicationState state, MethodInfo[] handlers, IntPtr appContext, HttpContext context) +169    System.Web.HttpApplicationFactory.GetSpecialApplicationInstance(IntPtr appContext, HttpContext context) +372    System.Web.Hosting.PipelineRuntime.InitializeApplication(IntPtr appContext) +329  [HttpException (0x80004005): Method not found: 'System.Collections.ObjectModel.Collection`1<System.Net.Http.DelegatingHandler> System.Web.Http.HttpConfiguration.get_MessageHandlers()'.]    System.Web.HttpRuntime.FirstRequestInit(HttpContext context) +525    System.Web.HttpRuntime.EnsureFirstRequestInit(HttpContext context) +118    System.Web.HttpRuntime.ProcessRequestNotificationPrivate(IIS7WorkerRequest wr, HttpContext context) +708
--

**Comment by dotnetjunkie** [comment]
Why do you think the problem lies with Simple Injector?

**Comment by dotnetjunkie** [comment]
But my guess is that your application (through the use of a library) references multiple versions of the System.Web.Http assembly (that contains the `System.Web.Http.HttpConfiguration` class), and that during startup or deployment the old version overrides the newer one that you used during compilation.

Make sure your application only references 1 specific version of System.Web.Http and make sure you have the proper binding redirects in place to make sure the assembly can be loaded.

**Comment by hemendpatel** [comment]
Thanks Steven - it was assembly version issue.

Now its working in first request and second request , I am getting exception

System.InvalidOperationException
  HResult=0x80131509
  Message=This instance has already started one or more requests. Properties can only be modified before sending the first request.
  Source=<Cannot evaluate the exception source>
  StackTrace:
<Cannot evaluate the exception stack trace>

 protected override Task<HttpResponseMessage> SendAsync(
            HttpRequestMessage request, CancellationToken cancellationToken)
        {

            // Important: Trigger the creation of the scope.
            request.GetDependencyScope();

            var handler = this.container.GetInstance<THandler>();

            **handler.InnerHandler = this.InnerHandler;**

            var invoker = new HttpMessageInvoker(handler);

            return invoker.SendAsync(request, cancellationToken);
        }

**Comment by dotnetjunkie** [comment]
Di you accidentally register your `RequestResponseLoggerHandler` as Singleton?

**Comment by hemendpatel** [comment]
thank you so much for your quick turn around. This really help me to meet my deadline.

I did it intestinally as I am getting below error and got reference on net to make it single ton. I was getting below error if I will not declare it as singleton.
The configuration is invalid. The following diagnostic warnings were reported:
-[Disposable Transient Component] HelloRequestHandler is registered as transient, but implements IDisposable.
See the Error property for detailed information about the warnings. Please see https://simpleinjector.org/diagnostics how to fix problems and how to suppress individual warnings.


**# Stack Trace:** 
[DiagnosticVerificationException:   The configuration is invalid. The following diagnostic warnings were   reported:   -[Disposable Transient   Component] HelloRequestHandler is registered as transient, but implements   IDisposable.   See the Error property for   detailed information about the warnings. Please see   https://simpleinjector.org/diagnostics how to fix problems and how to   suppress individual warnings.]        SimpleInjector.Container.ThrowOnDiagnosticWarnings() +225      SimpleInjector.Container.Verify() +61        SimpleInjector1.WebApiConfig.SimpleInjection(HttpConfiguration config)   in \SimpleInjector2\SimpleInjector1\App_Start\WebApiConfig.cs:41        SimpleInjector1.WebApiConfig.Register(HttpConfiguration config) in \SimpleInjector2\SimpleInjector1\App_Start\WebApiConfig.cs:27      System.Web.Http.GlobalConfiguration.Configure(Action`1   configurationCallback) +63        SimpleInjector1.WebApiApplication.Application_Start() in \SimpleInjector2\SimpleInjector1\Global.asax.cs:14       [HttpException (0x80004005):   The configuration is invalid. The following diagnostic warnings were   reported:   -[Disposable Transient   Component] HelloRequestHandler is registered as transient, but implements   IDisposable.   See the Error property for   detailed information about the warnings. Please see   https://simpleinjector.org/diagnostics how to fix problems and how to   suppress individual warnings.]        System.Web.HttpApplicationFactory.EnsureAppStartCalledForIntegratedMode(HttpContext   context, HttpApplication app) +535        System.Web.HttpApplication.RegisterEventSubscriptionsWithIIS(IntPtr   appContext, HttpContext context, MethodInfo[] handlers) +166        System.Web.HttpApplication.InitSpecial(HttpApplicationState state,   MethodInfo[] handlers, IntPtr appContext, HttpContext context) +169        System.Web.HttpApplicationFactory.GetSpecialApplicationInstance(IntPtr   appContext, HttpContext context) +372        System.Web.Hosting.PipelineRuntime.InitializeApplication(IntPtr   appContext) +329       [HttpException (0x80004005):   The configuration is invalid. The following diagnostic warnings were   reported:   -[Disposable Transient   Component] HelloRequestHandler is registered as transient, but implements   IDisposable.   See the Error property for   detailed information about the warnings. Please see   https://simpleinjector.org/diagnostics how to fix problems and how to   suppress individual warnings.]        System.Web.HttpRuntime.FirstRequestInit(HttpContext context) +525      System.Web.HttpRuntime.EnsureFirstRequestInit(HttpContext   context) +118        System.Web.HttpRuntime.ProcessRequestNotificationPrivate(IIS7WorkerRequest   wr, HttpContext context) +708
--

**Comment by dotnetjunkie** [comment]
Preferably, you shouldn't let your delegating handlers implement `IDisposable`. Unfortunately, there is no way around this, since delegating handlers inherit from `HttpMessageHandler`, which implements `IDisposable` :'(

DelegatingHandlers, however, are *not* thread-safe, and should not be registered as `Singleton`. The solution is to register them as `Scoped`. This ensures a `DelegatingHandler` does not outlive a single request, while ensuring the handler is disposed when the request ends.

Long story short, this is how to register your `RequestResponseLoggerHandler`:

``` c#
container.Register<RequestResponseLoggerHandler>(Lifestyle.Scoped);
```

**Comment by hemendpatel** [comment]
you are awesome. - got the root cause of it.

**Comment by GaneshSJ** [comment]
> Thanks Steven - it was assembly version issue.
> 
> Now its working in first request and second request , I am getting exception
> 
> System.InvalidOperationException
> HResult=0x80131509
> Message=This instance has already started one or more requests. Properties can only be modified before sending the first request.
> Source=
> StackTrace:
> ```cs
>  protected override Task<HttpResponseMessage> SendAsync(
>             HttpRequestMessage request, CancellationToken cancellationToken)
>         {
> 
>             // Important: Trigger the creation of the scope.
>             request.GetDependencyScope();
> 
>             var handler = this.container.GetInstance<THandler>();
> 
>             **handler.InnerHandler = this.InnerHandler;**
> 
>             var invoker = new HttpMessageInvoker(handler);
> 
>             return invoker.SendAsync(request, cancellationToken);
>         }
> ```
> 
> An error has occurred. This instance has already started one or more requests. Properties can only be modified before sending the first request. System.InvalidOperationException at System.Net.Http.DelegatingHandler.CheckDisposedOrStarted() at System.Net.Http.DelegatingHandler.set_InnerHandler(HttpMessageHandler value) at SimpleInjector1.Handler.DelegatingHandlerProxy`1.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken) in SimpleInjector2\SimpleInjector1\Handler\DelegatingHandlerProxy.cs:line 36 at System.Net.Http.DelegatingHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken) at System.Web.Http.HttpServer.d__24.MoveNext()

Hi hemendpatel,
can you please let us know in short how exactly you solved the issue? I am unable to point out the dll discrepancy on my machine, as there are several different versions of system.net.http, system.web, system.web.http, system.web.webpages.
Thanks

**Comment by luucongthin** [comment]
Hi hemendpatel,
can you please let us know in short how exactly you solved the issue? I am unable to point out the dll discrepancy on my machine, as there are several different versions of system.net.http, system.web, system.web.http, system.web.webpages.
Thanks
Hi hemendpatel,
can you please let us know in short how exactly you solved the issue? I am unable to point out the dll discrepancy on my machine, as there are several different versions of system.net.http, system.web, system.web.http, system.web.webpages.
Thanks

**Comment by luucongthin** [comment]
Hi hemendpatel,
can you please let us know in short how exactly you solved the issue? I am unable to point out the dll discrepancy on my machine, as there are several different versions of system.net.http, system.web, system.web.http, system.web.webpages.
Thanks
